### PR TITLE
Speed up HI miniAOD by removing unneeded features

### PR DESCRIPTION
**Pull request by mandrenguyen**
#### PR description:

Turns off several unused features in miniAOD:
- Disable pfPileUpIsoPFBRECO
- Remove algos from puppi, making puppi do nearly nothing
- no longer compute ghost tracks for impactParameterTagInfos
- Turn off area calculation for fat jet collection used for taus
CPU savings is around 10%
 
I also cleaned up pp_on_AA era statements in applySubstructure, which no longer do anything since applySubstructure is completely inactive for these eras

#### PR validation:

tested on 140.5611 and 300

### CODE REVIEW

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
this seems related to #31647; just to tag it here

**Comment by santocch** [comment]
+1

File: RecoBTag/ImpactParameter/python/impactParameterTagInfos_cfi.py [hunk 1]
```diff
@@ -19,4 +19,4 @@
 
 from Configuration.Eras.Modifier_pp_on_AA_2018_cff import pp_on_AA_2018
 from Configuration.Eras.Modifier_pp_on_PbPb_run3_cff import pp_on_PbPb_run3
-(pp_on_AA_2018 | pp_on_PbPb_run3).toModify(impactParameterTagInfos, jetTracks = "ak5JetTracksAssociatorAtVertex")
+(pp_on_AA_2018 | pp_on_PbPb_run3).toModify(impactParameterTagInfos, jetTracks = "ak5JetTracksAssociatorAtVertex", computeGhostTrack = False)
```
Comment (slava77) [inline RecoBTag/ImpactParameter/python/impactParameterTagInfos_cfi.py:22]: a line-break would be nice

File: RecoTauTag/Configuration/python/boostedHPSPFTaus_cff.py [hunk 2]
```diff
@@ -34,7 +34,7 @@
 from Configuration.Eras.Modifier_pp_on_XeXe_2017_cff import pp_on_XeXe_2017
 from Configuration.Eras.Modifier_pp_on_AA_2018_cff import pp_on_AA_2018
 from Configuration.Eras.Modifier_pp_on_PbPb_run3_cff import pp_on_PbPb_run3
-(pp_on_XeXe_2017 | pp_on_AA_2018 | pp_on_PbPb_run3).toModify(ca8PFJetsCHSprunedForBoostedTaus, inputEtMin = 999999.0, src = "particleFlow")
+(pp_on_XeXe_2017 | pp_on_AA_2018 | pp_on_PbPb_run3).toModify(ca8PFJetsCHSprunedForBoostedTaus, inputEtMin = 999999.0, src = "particleFlow", doAreaFastjet = False)
```
Comment (slava77) [inline RecoTauTag/Configuration/python/boostedHPSPFTaus_cff.py:37]: a line-break would be nice

**Review by slava77 — COMMENTED** [review]
mostly technical comments
the one for CommonTools/ParticleFlow/python/PFBRECO_cff.py is somewhat more significant than the rest

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
> Turns off several unused features in miniAOD:
> 
>     * Disable pfPileUpIsoPFBRECO

this is used to compute "PAT" level lepton isolation, saved in e.g. pat::Muons.
As a result of the update, the PU isolation sums of charged cands go to zero and the other PF isolations increase a bit (since PU subtraction is now doing nothing).

e.g. from 140.5611
all_test31983VSorig_RunHI2018reMINIAODwf140p5611c_patMuons_slimmedMuons__PAT_obj___isolations__4_
I noticed this while checking the execution time changes in 140.5611 and seeing that muon isolation-related module times go up by a factor of a few (about a half of the time gained in reducing/disabling `pfPileUpIsoPFBRECO`)

I guess from physics POV, disabling PU subtraction in HI events is likely the right thing to do. IIUC, the patMuons are going to be used in HI. Are these isolations not used?
If the intent is not to use them, then `toModify(patMuons, isolationValues = [])` may be a better choice. However, the update is probably doing the right thing from the first principles (at some extra CPU cost, of about 1.5% ).

@mandrenguyen 
please check and let me know what you want to do.

**Comment by mandrenguyen** [comment]
@slava77 This is not quite the behavior I expected.  Isolation is potentially useful, but it doesn't make sense to try to distinguish between charged hadrons from the hardest (or highest multiplicity) vertex and from PU.  The rate of real PU is small compared to the rate of mistakenly splitting the primary vertex.  So the desired behavior is simply to turn off CHS.  Can you advise how to do that?

**Comment by slava77** [comment]
> @slava77 This is not quite the behavior I expected. Isolation is potentially useful, but it doesn't make sense to try to distinguish between charged hadrons from the hardest (or highest multiplicity) vertex and from PU. The rate of real PU is small compared to the rate of mistakenly splitting the primary vertex. So the desired behavior is simply to turn off CHS. Can you advise how to do that?

From your comment it sounds like `toModify(patMuons, isolationValues = [])` is a bad choice and the preference would be to have an isolation with "CHS" disabled.

I think that what you did is the turning off of CHS for this specific isolation computation (`pfPileUpIsoPFBRECO, Enable = False`).

There are different variants of CHS (one for jets and one for isolations, both done via candidate "top projections"), these can be disabled in corresponding  `pfPileUp*, Enable = False`. Then, there is also a close equivalent to CHS in the packedPFCandidates production, done via `primaryVertexAssociation`), which ends up marked in `PackedCandidate::fromPV` and used in some isolation quantities as well (like muon miniIsolation); I'm not sure of the best way to "disable" this one. Oh, and then there is a PUPPI-based PU subtraction (but we disabled it for HI as it doesn't make as much sense).

Note, however, that this PR affects the isolations recomputed in PAT and saved in `pat::Muon::isolations_`; if I'm not mistaken they are not the main isolations used in pp.
The more "standard" muon isolations are in the pat::Muons, inherited and computed for reco::Muon in `pfIsolationR03` or R04. These were made in AOD and used the isolation CHS. Perhaps the easiest way to deal with this is at analysis level and add `sumPUPt` to the `sumChargedHadronPt` (or/and not to use the "delta-beta" prescriptions for pilup corrections). 

BTW, things get further complicated by the layer of PF cleaning that you added in the HI miniAOD. If you actually want this applied to the muon isolations, I guess you may then be better off to plan to use the `pat::Muon::isolations_` (modified by this PR already), but the input collections to `*IsoPFBRECO` will need to be changed to `cleanedParticleFlow`.

So,  not a simple answer, and it mostly covers just muons.

**Comment by mandrenguyen** [comment]
@slava77 I propose to just roll back the change to  CommonTools/ParticleFlow/python/PFBRECO_cff.py
It sounds like the muon isolation is not going to be usable out of the box, so I propose we not modify the default settings.

**Comment by slava77** [comment]
> @slava77 I propose to just roll back the change to CommonTools/ParticleFlow/python/PFBRECO_cff.py
> It sounds like the muon isolation is not going to be usable out of the box, so I propose we not modify the default settings.

OK, even though the choice is not particularly obvious 

(offhand, I would have even though that the current variant is even more appropriate; but that's more of a guess implying some studies would still be good to confirm the choices).  It sounds like it's more likely that you will pick the right (optimal) isolation based on recomputation using existing packed cands.

**Comment by mandrenguyen** [comment]
@slava77 Indeed, I think we would have to recompute this.  Isolation is used infrequently enough that I don't think this is a huge drawback.  Perhaps we can rethink this for Run 3.  Since disabling this feature is a bit faster, let's just keep the PR as it is then.

**Comment by slava77** [comment]
+1

for 
- code changes are in line with the PR description and the follow up discussion
- jenkins tests pass and comparisons with the baseline show differences only in HI workflow 158.0 where the PUPPI weights are now seen as zeroed out
- local test with 140.5611 confirms about 10% reduction in CPU time and shows more differences in the isolation variables recomputed at PAT level
    - lepton PUPPI-based isolation is now at zero, as expected from disabling puppi (expected behavior)
    - the `PFBRECO` flavor of isolations was discussed earlier in the thread and the change is taken as somewhat OK, with a quite likely case that an isolation suitable for analysis can be recomputed in analysis

**Comment by slava77** [comment]
@qliphy @silviodonato 
may I suggest not to wait for analysis signature here.
Please consider.
Thank you.

**Comment by qliphy** [comment]
@mandrenguyen   after merging #31935, there are conflicts to be fixed for this PR.  Please have a check.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/31893	8c48ced 
- after the rebase the updates which were previously applied to `pp_on_AA_2018 | pp_on_PbPb_run3` now consistently apply to `pp_on_AA`, as expected from the refactoring of modifiers introduced by  #31935
- results of jenkins tests are the same as in the previous version

**Comment by santocch** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**